### PR TITLE
Missing models to fix tests

### PR DESCRIPTION
**Pull request by FabrizioCaldarelli**
Missing models to fix tests

### CODE REVIEW

File: tests/data/Animal.php [hunk 1]
```diff
@@ -23,9 +23,13 @@ public static function tableName()
         return 'animal';
     }
 
+    public function __construct(array $config = [])
+    {
+        $this->init();
```
Comment (machour) [inline tests/data/Animal.php]: As a rule of thumb, `init()` should never be explicitly calling.

It was the old way Yii was doing things internally.

We can either switch the class to use the new DI mechanism, or implement \yii\di\Initiable in the class, and the Factory will take care of calling init() for you. 

Here, I think you can just move `$this->type = \get_called_class();` in the constructor and remove init() completely.

File: src/AfterSaveEvent.php [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * @author Carsten Brandt <mail@cebe.cc>
+ * @since 2.0
+ */
+class AfterSaveEvent extends Event
+{
+    /**
+     * @var array The attribute values that had changed and were saved.
+     */
+    public $changedAttributes;
```
Comment (machour) [inline src/AfterSaveEvent.php]: I think that Events have been completely reworked in Yii 3, and this Event was renamed.
Check \yii\base\ValidationEvent for example.

File: src/ActiveQuery.php [hunk 3]
```diff
@@ -106,9 +107,13 @@ class ActiveQuery extends Query implements ActiveQueryInterface
...
      * @param string $modelClass the model class associated with this query
      */
-    public function __construct($modelClass)
+    public function __construct($modelClass, array $config = [])
     {
         $this->modelClass = $modelClass;
+
+        AbstractContainer::configure($this, $config);
+
+        $this->init();
```
Comment (machour) [inline src/ActiveQuery.php]: see remark below

File: tests/data/Animal.php (hunk 1 above)
Comment (FabrizioCaldarelli) [reply tests/data/Animal.php]: Ok done

File: src/AfterSaveEvent.php (hunk 2 above)
Comment (FabrizioCaldarelli) [reply src/AfterSaveEvent.php]: How does yii\base\ValidationEvent support changedAttributes?

File: src/ActiveQuery.php (hunk 3 above)
Comment (FabrizioCaldarelli) [reply src/ActiveQuery.php]: Done

**Comment by FabrizioCaldarelli** [comment]
Are there other issues about this PR ?

File: tests/data/Animal.php [hunk 4]
```diff
@@ -23,9 +23,8 @@ public static function tableName()
         return 'animal';
     }
 
-    public function init()
+    public function __construct(array $config = [])
```
Comment (machour) [inline tests/data/Animal.php]: $config is not needed, you can remove it

Comment (FabrizioCaldarelli) [reply tests/data/Animal.php]: Removed

Comment (terabytesoftw) [reply tests/data/Animal.php]: in BaseActiveRecord.php:

You should not change AfterSaveEvent for ValidationEvent.

```
public function afterSave ($ insert, $ changedAttributes)
{
    $ this-> trigger (new ValidationEvent (
        $ insert? self :: EVENT_AFTER_INSERT: self :: EVENT_AFTER_UPDATE,
        $ changedAttributes
     ));
}
```

File: src/ActiveQuery.php [hunk 5]
```diff
@@ -117,7 +121,7 @@ public function __construct($modelClass)
      * an [[EVENT_INIT]] event. If you override this method, make sure you call the parent implementation at the end
      * to ensure triggering of the event.
      */
-    public function init()
+    public function init() : void
     {
         $this->trigger(self::EVENT_INIT);
```
Comment (hiqsol) [inline src/ActiveQuery.php]: Events were redone.
Please see `yii\widget\Widget::init()` to see example.

File: src/BaseActiveRecord.php [hunk 6]
```diff
@@ -969,7 +969,7 @@ public function afterFind()
      */
     public function beforeSave($insert)
     {
-        $event = new ModelEvent(['name' => $insert ? self::EVENT_BEFORE_INSERT : self::EVENT_BEFORE_UPDATE]);
+        $event = new ValidationEvent( $insert ? self::EVENT_BEFORE_INSERT : self::EVENT_BEFORE_UPDATE );
```
Comment (hiqsol) [inline src/BaseActiveRecord.php]: Please look here for example:
https://github.com/yiisoft/yii-core/blob/1e4f498e52c630f55ee06e4d949bb5ed4ed37c98/src/base/Model.php#L381

File: src/BaseActiveRecord.php [hunk 7]
```diff
@@ -995,10 +995,10 @@ public function beforeSave($insert)
...
     public function afterSave($insert, $changedAttributes)
     {
-        $this->trigger(new AfterSaveEvent([
-            'name' => $insert ? self::EVENT_AFTER_INSERT : self::EVENT_AFTER_UPDATE,
-            'changedAttributes' => $changedAttributes,
-        ]));
+        $this->trigger(new AfterSaveEvent(
+            $insert ? self::EVENT_AFTER_INSERT : self::EVENT_AFTER_UPDATE,
+            $changedAttributes
+        ));
```
Comment (hiqsol) [inline src/BaseActiveRecord.php]: Please see similar example here:
https://github.com/yiisoft/yii-core/blob/1e4f498e52c630f55ee06e4d949bb5ed4ed37c98/src/base/Controller.php#L276
It could be done like this:
`$this->trigger(AfterSaveEvent::create($insert, $changedAttributes)`

File: src/ActiveQuery.php (hunk 5 above)
Comment (FabrizioCaldarelli) [reply src/ActiveQuery.php]: This is widget init(): 

```
    public function init(): void
    {
        $this->trigger(WidgetEvent::INIT);
    }
```

Should we create an ActiveQueryEvent class? If so, it should be like as:

```
class ActiveQueryEvent extends Event
{
    public const INIT = 'activeQuery.init';
}

```

Is it right?

File: src/BaseActiveRecord.php (hunk 6 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: So, it should be:

>         return $this->trigger($insert ? self::EVENT_BEFORE_INSERT : self::EVENT_BEFORE_UPDATE);

In your example, trigger parameter is from ValidationEvent class. Have we a class to get EVENT_BEFORE_INSERT and EVENT_BEFORE_UPDATE?

Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: I've created a new ActiveRecordSaveEvent: https://github.com/yiisoft/active-record/blob/31ab3481a0468a9cf47f4621319f6c8bbf5a5ea9/src/ActiveRecordSaveEvent.php

I was thinking to create a single ActiveRecordEvent to collect all events related to ActiveRecord: find, save, delete, etc.. etc.., instead a single *Event for each of them.

The reason is that, for example, beforeSave and afterSave pass $insert and $changedAttributes parameters, that they should be shared with other events (find, delete, etc.).

File: src/BaseActiveRecord.php (hunk 7 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: I've created a new ActiveRecordSaveEvent: https://github.com/yiisoft/active-record/blob/31ab3481a0468a9cf47f4621319f6c8bbf5a5ea9/src/ActiveRecordSaveEvent.php

I was thinking to create a single ActiveRecordEvent to collect all events related to ActiveRecord: find, save, delete, etc.. etc.., instead a single *Event for each of them.

The reason is that, for example, beforeSave and afterSave pass $insert and $changedAttributes parameters, that they should be shared with other events (find, delete, etc.).

**Comment by machour** [comment]
Just wondering if we should create a `src/events/` folder and host events under `\yii\activerecord\events\`.

We could also remove 'ActiveRecord' from events name since we're already in yiisoft/activerecord

**Comment by FabrizioCaldarelli** [comment]
I was thinking the same.

File: src/ActiveRecordDeleteEvent.php [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+ * @author Fabrizio Caldarelli <fabrizio.caldarelli@gmail.com>
+ * @since 3.0
+ */
+class ActiveRecordDeleteEvent extends Event
+{
+    /**
+     * @event event raised at the beginning of [[delete()]]. You may set
+     * [[Event::isValid]] to be false to stop the validation.
+     */    
+    const BEFORE = 'activeRecordDelete.before';
```
Comment (machour) [inline src/ActiveRecordDeleteEvent.php]: Naming could be better like this: `active-record.delete.before`
(snake case with '-', and a '.' between every significant verb)

File: src/BaseActiveRecord.php [hunk 9]
```diff
@@ -11,11 +11,13 @@
 use yii\base\InvalidCallException;
 use yii\base\InvalidConfigException;
 use yii\base\Model;
-use yii\base\ModelEvent;
+use yii\base\ValidationEvent;
```
Comment (machour) [inline src/BaseActiveRecord.php]: not needed I think

File: src/BaseActiveRecord.php [hunk 10]
```diff
@@ -929,7 +890,7 @@ public function setIsNewRecord($value)
     public function init()
     {
         parent::init();
-        $this->trigger(self::EVENT_INIT);
+        $this->trigger(ActiveRecordInitEvent::init());
```
Comment (machour) [inline src/BaseActiveRecord.php]: Remove the `ActiveRecord` prefix from classes names

File: src/BaseActiveRecord.php [hunk 11]
```diff
@@ -940,7 +901,7 @@ public function init()
      */
     public function afterFind()
     {
-        $this->trigger(self::EVENT_AFTER_FIND);
+        $this->trigger(ActiveRecordFindEvent::after());
```
Comment (machour) [inline src/BaseActiveRecord.php]: I'm not sure why most calls are not  `Class::AFTER` instead of the static function call 🤔 
Could you enlighten me?

File: tests/unit/ActiveRecordTestTrait.php [hunk 12]
```diff
@@ -1102,9 +1104,9 @@ public function testAfterFind()
         /* @var $this TestCase|ActiveRecordTestTrait */
 
         $afterFindCalls = [];
-        Event::on(BaseActiveRecord::className(), BaseActiveRecord::EVENT_AFTER_FIND, function ($event) use (&$afterFindCalls) {
+        Event::on('\yii\activerecord\BaseActiveRecord', ActiveRecordFindEvent::AFTER, function ($event) use (&$afterFindCalls) {
```
Comment (machour) [inline tests/unit/ActiveRecordTestTrait.php]: I think that using `::class` instead of strings is better (less error prone)

File: src/BaseActiveRecord.php (hunk 10 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: If we use only InitEvent as name for event class, we could create a mess, for example for ActiveRecordInitEvent and ActiveQueryInitEvent, or between same event type but different base name classes.

File: src/BaseActiveRecord.php (hunk 11 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: In ActiveRecordSaveEvent, we need to pass extra data ( $insert and $changedAttributes ) to the event object.

https://github.com/FabrizioCaldarelli/active-record/blob/missing-models/src/ActiveRecordSaveEvent.php

So I would uniform syntax to trigger event, otherwise we could have some cases with ::CLASS and other with ClassName::eventName( args )

File: src/ActiveRecordDeleteEvent.php (hunk 8 above)
Comment (machour) [reply src/ActiveRecordDeleteEvent.php]: @rob006 any suggestion?

File: src/BaseActiveRecord.php (hunk 10 above)
Comment (machour) [reply src/BaseActiveRecord.php]: Agreed, I didn't see that there were also ActiveQueryXXX events

File: src/BaseActiveRecord.php (hunk 11 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: Is not more descriptive ::after() instead new ActiveRecordFindEvent('after', ....)?

Assuming that I should pass as first parameter the action.

File: tests/unit/ActiveRecordTestTrait.php (hunk 12 above)
Comment (FabrizioCaldarelli) [reply tests/unit/ActiveRecordTestTrait.php]: I'm agree

File: src/ActiveRecordDeleteEvent.php (hunk 8 above)
Comment (rob006) [reply src/ActiveRecordDeleteEvent.php]: @machour I would leave it as is or use FQN (like `yii.activerecord.ActiveRecordDeleteEvent:before`).

Comment (FabrizioCaldarelli) [reply src/ActiveRecordDeleteEvent.php]: Since we are at the beginning and changes are easily to do, I prefer FQN

File: src/BaseActiveRecord.php (hunk 11 above)
Comment (rob006) [reply src/BaseActiveRecord.php]: > Is not more descriptive ::after() instead new ActiveRecordFindEvent('after', ....)?

Not really. For me, it just hides what is really happening and how `trigger()` works. 

And I'm not really sure about using one class for different events. It may be hard to follow this consistently - what if `after` event will need different properties than `before` event? IMO it breaks SRP.

Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: If you need different properties, you can pass them to the method. What is the problem?

Comment (rob006) [reply src/BaseActiveRecord.php]: But why I need separate method as a wrapper for `new` keyword? It is less readable and less flexible (I can't override event and add required parameter to the function, while I can do this in constructor).

Comment (hiqsol) [reply src/BaseActiveRecord.php]: This approach is called "named constructor" (or static factory) and it is indeed more readable (because it is shorter and more expressive) and more flexible cause it allows rewriting event class and actual constructor while keeping "named constructors" signatures.

File: src/ActiveRecordSaveEvent.php [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        $this->_insert = $insert;
+        $this->_changedAttributes = $changedAttributes;
+    }
+
+    /**
+     * Creates BEFORE SAVE event.
+     * @param string $insert specify if action is insert or update ( true for insert, false for update )
+     * @return self created event
+     */
+    public static function before($insert): self
```
Comment (hiqsol) [inline src/ActiveRecordSaveEvent.php]: Add type hinting, please.

File: src/ActiveRecordSaveEvent.php [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        return new static(static::BEFORE, $insert);
+    }
+
+    /**
+     * Creates AFTER SAVE event.
+     * @param string $insert specify if action is insert or update ( true for insert, false for update )
+     * @param string $changedAttributes list of changed attributes
+     * @return self created event
+     */
+    public static function after($insert, $changedAttributes): self
```
Comment (hiqsol) [inline src/ActiveRecordSaveEvent.php]: Type hinting too, please.

File: src/ActiveRecordSaveEvent.php [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        return (new static(static::AFTER, $insert, $changedAttributes));
+    }
+
+    public function getInsert() : bool { return $this->_insert; }
```
Comment (hiqsol) [inline src/ActiveRecordSaveEvent.php]: Format PSR-2, please

Comment (FabrizioCaldarelli) [reply src/ActiveRecordSaveEvent.php]: Done

File: src/ActiveRecordSaveEvent.php (hunk 14 above)
Comment (FabrizioCaldarelli) [reply src/ActiveRecordSaveEvent.php]: Done

File: src/ActiveRecordSaveEvent.php (hunk 13 above)
Comment (FabrizioCaldarelli) [reply src/ActiveRecordSaveEvent.php]: Done

File: tests/unit/ActiveRecordTestTrait.php (hunk 12 above)
Comment (FabrizioCaldarelli) [reply tests/unit/ActiveRecordTestTrait.php]: Done

File: src/BaseActiveRecord.php (hunk 11 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: Done

File: src/BaseActiveRecord.php (hunk 9 above)
Comment (FabrizioCaldarelli) [reply src/BaseActiveRecord.php]: Done

File: src/ActiveRecordDeleteEvent.php (hunk 8 above)
Comment (FabrizioCaldarelli) [reply src/ActiveRecordDeleteEvent.php]: used full namespace

**Comment by machour** [comment]
Awesome, thank you! 🙌
